pkg/operator/controller/ingress_status.go

```diff
@@ -101,3 +139,148 @@ func ingressStatusesEqual(a, b operatorv1.IngressControllerStatus) bool {
 
 	return true
 }
+
+func isProvisioned(service *corev1.Service) bool {
+	ingresses := service.Status.LoadBalancer.Ingress
+	return len(ingresses) > 0 && (len(ingresses[0].Hostname) > 0 || len(ingresses[0].IP) > 0)
+}
+
+func isPending(service *corev1.Service) bool {
+	return !isProvisioned(service)
+}
+
+func controlledBy(service *corev1.Service, matches func(*corev1.Service) bool) []string {
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return []string{}
+	}
+	if !matches(service) {
+		return []string{}
+	}
+	controller, ok := service.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		return []string{}
+	}
+	return []string{controller}
+}
+
+func indexByPublishing(obj runtime.Object) []string {
+	c, ok := obj.(*operatorv1.IngressController)
+	if !ok {
+		return []string{}
+	}
+	return []string{string(c.Status.EndpointPublishingStrategy.Type)}
+}
+
+func indexByLoadBalancerPublishing(obj runtime.Object) []string {
+	c, ok := obj.(*operatorv1.IngressController)
+	if !ok {
+		return []string{}
+	}
+	if c.Status.EndpointPublishingStrategy.Type == operatorv1.LoadBalancerServiceStrategyType {
+		return []string{c.Name}
+	}
+	return []string{}
+}
+
+func indexByProvisioned(obj runtime.Object) []string {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}
+	}
+	return controlledBy(service, isProvisioned)
+}
+
+func indexByPending(obj runtime.Object) []string {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}
+	}
+	return controlledBy(service, isPending)
+}
+
+var serviceIndexers = map[string]client.IndexerFunc{
+	"pending-for":     indexByPending,
+	"provisioned-for": indexByProvisioned,
+}
+var controllerIndexers = map[string]client.IndexerFunc{
+	"publishing":          indexByPublishing,
+	"wants-load-balancer": indexByLoadBalancerPublishing,
+}
+
+type ingressStatusCache struct {
+	contains func(kind runtime.Object, name, value string) bool
+}
+
+func cacheContains(cache cache.Cache, list runtime.Object, name, value string) bool {
+	err := cache.List(context.TODO(), list, client.MatchingField(name, value))
+	if err != nil {
+		return false
+	}
+	// TODO: after rebase, replace with:
+	// meta.LenList(list) > 0
+	items, _ := meta.ExtractList(list)
+	return len(items) > 0
+}
+
+func NewIngressStatusCache(c cache.Cache) *ingressStatusCache {
+	add := func(cache cache.Cache, kind runtime.Object, indexers map[string]client.IndexerFunc) {
+		for name := range indexers {
+			cache.IndexField(kind, name, indexers[name])
```

Comment:
The error return value is ignored.

Reply:
If we get an error here I think we should probably panic anyway? It would mean we wired up everything wrong from the start